dev/buildtool/source_code_manager.py

```diff
@@ -0,0 +1,294 @@
+    logging.debug('Loading bom version %s', version)
+    bom_yaml_string = check_subprocess(
+        'hal version bom {0} --color false --quiet'
+        .format(version), echo=False)
+    return yaml.load(bom_yaml_string)
+
+  def maybe_pull_repository_source(
+      self, repository,
+      bom_version=None, bom_path=None,
+      git_branch=None, default_branch=None):
+    """Pull the source as specified, if it was specified."""
+    # pylint: disable=too-many-arguments
+
+    pull_version = 1 if bom_version else 0
+    pull_path = 1 if bom_path else 0
+    pull_branch = 1 if git_branch else 0
+    have = pull_version + pull_path + pull_branch
+    if have == 0:
+      logging.debug('No source pull requests to process.')
+      return
+
+    if have > 1:
+      raise ValueError('Ambguous source code requests.')
+
+    if default_branch and not git_branch:
+      raise ValueError('A default_branch requires a git_branch')
+
+    git_dir = self.get_local_repository_path(repository.name)
+    if os.path.exists(git_dir):
+      message = '%s already exists. Cannot pull source.' % git_dir
+      logging.error(message)
+      raise ValueError(message)
+
+    logging.info('Pulling %s to %s', repository.url, git_dir)
+    ensure_dir_exists(os.path.dirname(git_dir))
+
+    if git_branch:
+      self.git.clone_repository_to_path(
+          repository.url, git_dir,
+          branch=git_branch, default_branch=default_branch)
+    else:
+      if bom_version:
+        bom = self.bom_from_version(bom_version)
+      else:
+        bom = self.bom_from_path(bom_path)
+      self.pull_source_from_bom(repository.name, git_dir, bom)
+
+  def pull_source_from_bom(self, repository_name, git_dir, bom):
+    """Pull the source for a particular repository name from a bom."""
+    if repository_name in ('monitoring-daemon', 'monitoring-third-party'):
+      repository_name = 'spinnaker-monitoring'
+
+    spec = bom['services'][repository_name]
+    commit = spec['commit']
+    version = spec['version']
+
+    git_tag = 'version-{}'.format(version[:version.index('-')])
+
+    # We're creating a local repo containing only the commit and tag.
+    # The reason for this is because nebula gets confused by our tags
+    # and wants to only use the latest Netflix tag, which is wrong.
+    # If we could control nebula, we wouldnt need to wipe the repo and
+    # add a new one, but we'd still need to add the new tag to ensure it
+    # exists because the bom may not have been pushed.
+    if os.path.exists(git_dir):
+      return None
+
+    base_url = bom['artifactSources']['gitPrefix']
+    origin_url = spec.get(
+        'sourceRepository',
+        '{base}/{name}'.format(base=base_url, name=repository_name))
+    branch = spec.get(
+        'sourceBranch',
+        bom['artifactSources'].get('gitBranch', '(unknown)'))
+
+    self.__git.clone_repository_to_path(origin_url, git_dir, commit=commit)
+    self.__git.reinit_local_repository_with_tag(
```

Comment:
Looks like we are re-initializing with the git repo with a `version-{}` tag before we run the gradle builds. What are the are the resulting versions for the debians and container images - do they contain `version-`?

Reply:
If you are looking at this, I assume you are choosing the review it all at once option.
I thought it might be easier to digest in pieces, but your call (sorry brandon -- we can do a separate walk through)

This part is a little weird because of nebula. I thought that the repo needs to be tagged before gradle because gradle uses the tag. It is tagged with the version.

> git -C scratch/build/clouddriver show-ref --tags
d6887d526968ef544ca4c205fa9e716607fa1a5a refs/tags/2.0.0-20171218175040

However I also record what I want the "version" tag to be for later
> cat scratch/clouddriver/clouddriver-summary.yml
commit_id: 623e2457708e9f2b86dfe45811c22016abb572f5
tag: version-2.0.0
version: 2.0.0

The artifact gradle publishes to bintray is this:
About This Version
Website
https://home/ewiseblatt/build/newtest/clouddriver
Issue Tracker
https://home/ewiseblatt/build/newtest/clouddriver
VCS
https://home/ewiseblatt/build/newtest/clouddriver.git
Licenses
Apache-2.0
VCS Tag
v2.0.0-20171218175040
Debian Details
spinnaker-clouddriver_2.0.0-20171218175040_all.deb
Name	spinnaker-clouddriver
Version	2.0.0-20171218175040
Indexed Architecture	i386, amd64
Indexed Component	spinnaker
Indexed Distribution	xenial, trusty ControlFileContent:	
Package	spinnaker-clouddriver
Source	spinnaker-clouddriver
Version	2.0.0-20171218175040
Section	java
Priority	optional
Architecture	all
Installed-Size	180375
Maintainer	ewiseblatt
Description	spinnaker-clouddriver clouddriver-web


The sequencing is a bit different from the original scripts, but I dont have to re-pull anything or modify the local repository associated with the upstream/origin that we'll be pushing tags to later. When I had a "version-" tag in the gradle repo, I was getting something weird that caused parsing errors in gradle if I recall.

The scratch/<component> contains the intermediate state information passed among commands, which are now executing in different processes so dont have shared memory. But before were being regenerated anyway, so was confusing and not as clearly connected to where and when the version information was being computed.

The scratch/build/<component> is the munged directory to satisfy nebula. If/when nebula is fixed then this can go away and the abstraction maintaining it can be changed to refer to the <root_dir>/<component> repository.

BTW, the example above is version 2.0 because of this:
D 17:50:41.950 [Thread-5.23936] Commit is considered "MAJOR" because it says "('BREAKING CHANGE: Stops support of CF until maintainer is found',)"

Reply:
Just had a question on the usage of this specific function, would still like to review this in pieces :)

Thanks for the explanation.